src/hotspot/share/gc/g1/g1ParScanThreadState.cpp

```diff
@@ -623,15 +623,14 @@ oop G1ParScanThreadState::handle_evacuation_failure_par(oop old, markWord m, siz
     // Forward-to-self succeeded. We are the "owner" of the object.
     HeapRegion* r = _g1h->heap_region_containing(old);
 
-    // Objects failing evacuation will turn into old objects since the regions
-    // are relabeled as such. We mark the failing objects in the marking bitmap
-    // and later use it to handle all failed objects.
-    _g1h->mark_evac_failure_object(old);
-
     if (_evac_failure_regions->record(r->hrm_index())) {
       _g1h->hr_printer()->evac_failure(r);
     }
 
+    // Mark the failing object in the marking bitmap and later use the bitmap to handle
+    // evacuation failure recovery.
+    _g1h->mark_evac_failure_object(_worker_id, old, word_sz);
```

Comment:
The same live-size is calculated again in `process_chunk`, isn't it?

Reply:
You mean the one (ultimately) recorded in the `G1ConcurrentMark::_live_bytes` array and the one for the statistics in `G1RemoveSelfForwardsInChunksTask::process_chunk`?
You probably suggest to merge them somehow to do this work once too?

I do not think this is worth, particularly because `process_chunk` only collects live bytes per thread without any synchronization (it does collect dead bytes with synchronization).
So you could use that value (the total dead bytes) or vice versa to update `G1ConcurrentMark::_live_bytes` at the end of collection, but
* `G1ConcurrentMark::_live_bytes` (the sum) isn't actually gathered here, but again these are basically per-thread values (using the mark-stats cache), so otoh the update is inexpensive (nothing compared to the `par_mark` in `mark_evac_failure_object`), and otoh it can't be used for updating `HeapRegion::_garbage_bytes` directly. This would need to flush buffers somewhere before doing that.
* the collected `HeapRegion::_garbage_bytes` could be used at the end to update `G1ConcurrentMark::_live_bytes`, but it does not seem to be worth the complexity adding another phase somewhere else. Also it would require another `G1ConcurrentMark` API entry just for that which I do not see worth doing.

I.e. I do not think it is worth optimizing something that trivial and add the required code complexity. I would prefer to decrease code complexity.

Reply:
I was thinking sth like:
```
void HeapRegion::note_self_forwarding_removal_end_par(size_t garbage_bytes) {
  Atomic::store(&_live_bytes, region_size - _garbage_bytes, ...);
  Atomic::add(&_garbage_bytes, garbage_bytes, memory_order_relaxed);
}
```

Reply:
I see; I was assuming `_live_bytes` and `_garbage_bytes` live in the same class/context/entity. In current impl, `_live_bytes` lives in CM while `_garbage_bytes` in `HeapRegion`, which is quite confusing IMO.

Reply:
There has never been a `_live_bytes` as implied by the name anywhere (there is a getter on `HeapRegion` that (still) calculates an upper bound on the number of live bytes in a region; it uses `_garbage_bytes` to determine that). The use of `_live_bytes` is maybe a misnomer in the marking code (but for marking only something below TAMS is relevant).

I am open to renaming it (separately).

In `HeapRegion` the name of this counter has always been `_marked_bytes` which meant "marked bytes below TAMS". TAMS is something that is only useful (and valid) during marking, a `_marked_bytes` on `HeapRegion` is imho very questionable (because `tams == bottom` most of the time). I.e. having nonzero marked bytes in `[bottom, tams[`  where `tams == bottom` is even more inconsistent.

Marked bytes for a region only has meaning with an associated TAMS. That's why it's gone, and it is unnecessary to store in the first place as the removal change showed.

Garbage bytes is the lower bound of dead objects between [bottom, top[, which has meaning regardless of other temporary marking data like TAMS.

This is one reason I am suggesting to (also) move TAMS out of `HeapRegion`. I only wanted to wait with that until after this change; another that TAMS is intrinsically linked to the bitmap (and marking) so it should be managed (like TARS) by `G1ConcurrentMark`, and hence cleared with the bitmap in one go.

That G1 messes with the bitmap for evacuation failure handling is from an abstraction POV not good, but overall acceptable.

Maybe the `clear_bitmap_for_region()` method should (already) be called something like `clear_marking_data_in_region()`?